File: frontend/src/services/queryService.js
import api from './api';

/**
 * 查询服务 - 处理查询相关的API调用
 */
const queryService = {
  /**
   * 执行查询
   * @param {Object} queryData - 查询数据
   * @param {string} queryData.question - 查询问题
   * @param {string} queryData.mode - 查询模式 (AUTO/NLP/RAG)
   * @returns {Promise} 查询结果
   */
  query: async (queryData) => {
    return api.post('/query', queryData);
  },

  // 别名方法，保持向后兼容
  executeQuery: async (queryData) => {
    return queryService.query(queryData);
  },

  /**
   * 获取查询历史
   * @param {number} limit - 返回记录数量限制
   * @returns {Promise} 查询历史列表
   */
  getHistory: async (limit = 20) => {
    return api.get('/history', {
      params: { size: limit },
    });
  },

  /**
   * 获取查询历史（分页）
   * @param {number} page - 页码
   * @param {number} size - 每页大小
   * @returns {Promise} 查询历史列表
   */
  getQueryHistory: async (page = 0, size = 20) => {
    return api.get('/history', {
      params: { page, size },
    });
  },

  /**
   * 根据ID获取查询历史详情
   * @param {number} id - 查询历史ID
   * @returns {Promise} 查询历史详情
   */
  getQueryHistoryById: async (id) => {
    return api.get(`/history/${id}`);
  },

  /**
   * 删除查询历史
   * @param {number} id - 查询历史ID
   * @returns {Promise} 删除结果
   */
  deleteQueryHistory: async (id) => {
    return api.delete(`/history/${id}`);
  },

  /**
   * 清空所有查询历史
   * @returns {Promise} 清空结果
   */
  clearAllHistory: async () => {
    return api.delete('/history/all');
  },
};

export default queryService;

File: backend/src/main/java/com/scb/ragdemo/service/QueryService.java
package com.scb.ragdemo.service;

import com.scb.ragdemo.model.dto.request.QueryRequest;
import com.scb.ragdemo.model.dto.response.QueryResponse;
import com.scb.ragdemo.model.entity.QueryHistory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Query Service Interface
 * Responsible for handling user query requests and intelligent routing
 */
public interface QueryService {
    
    /**
     * Process user query request
     * Automatically selects NLP or RAG mode based on the request
     * 
     * @param request Query request object
     * @return Query response result
     */
    QueryResponse processQuery(QueryRequest request);
    
    /**
     * Execute query using NLP mode
     * Directly calls Poe API
     * 
     * @param question User question
     * @return Query response result
     */
    QueryResponse queryWithNlp(String question);
    
    /**
     * Execute query using RAG mode
     * Retrieves relevant document chunks and combines them with the question to call Poe API
     * 
     * @param question User question
     * @param topK Number of top relevant document chunks to retrieve
     * @return Query response result
     */
    QueryResponse queryWithRag(String question, int topK);
    
    /**
     * Get query history with pagination support
     * 
     * @param pageable Pagination parameters
     * @return Query history paginated result
     */
    Page<QueryHistory> getQueryHistory(Pageable pageable);
    
    /**
     * Get query history filtered by mode with pagination
     * 
     * @param mode Query mode
     * @param pageable Pagination parameters
     * @return Query history paginated result
     */
    Page<QueryHistory> getQueryHistoryByMode(String mode, Pageable pageable);
    
    /**
     * Delete specified query history record
     * 
     * @param id Query history record ID
     */
    void deleteQueryHistory(Long id);
    
    /**
     * Clear all query history
     */
    void clearAllHistory();
    
    /**
     * Get query statistics information
     * 
     * @return Statistics information Map
     */
    java.util.Map<String, Object> getQueryStatistics();
    
    /**
     * Get cached query result
     * 
     * @param question User question
     * @param mode Query mode
     * @return Cached result if exists, otherwise null
     */
    QueryResponse getCachedResult(String question, String mode);
    
    /**
     * Cache query result
     * 
     * @param question User question
     * @param mode Query mode
     * @param response Query response result
     */
    void cacheResult(String question, String mode, QueryResponse response);
}
